File: Problem #3/problem3_solution.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

tree = ""
def serialize(node, tree=""):
    """Serialize a binary tree into a string 
    Since it is a binary tree, the tree can have empty (Null) children. Mark these with '#'.

    Parameters: 
    argument1 Node: A node containing either children or nothing (Null)

    Return:
    String: The node serialized into a string format
    """
    
  
    if (not node): #Base case
        tree += "# "
        return tree
    tree += (str(node.val) + " ")
    tree = serialize(node.left, tree)
    tree = serialize(node.right, tree)

    return tree

i = 0
def deserialize(stringTree):
    """Deserialize a string into a binary tree. 
    Nodes with empty children (Null) are marked with #

    Parameters: 
    argument1 s: A list of strings containing the tree

    Return:
    Node: A binary tree 
    """
    global i #Global shared index 

    if (stringTree[i] == "#"): #Empty node
        return None 

    else:
        node = Node(stringTree[i]) #Create node of current index
        i+=1
        node.left = deserialize(stringTree) #Check left child
        i+=1
        node.right = deserialize(stringTree) #Check right child
        return node


node = Node('root', Node('left', Node('left.left')), Node('right'))

assert deserialize(serialize(node).split()).left.left.val == 'left.left'

#Original
#assert deserialize(serialize(node)).left.left.val == 'left.left'

File: Problem #3/problem3_solution.js
class Node {
    constructor(val, left, right) {
        this.val = val
        this.left = left
        this.right = right
    }
}

/**
 * Serialize a binary tree into a string 
 * Since it is a binary tree, the tree can have empty (Null) children. Mark these with '#'.
 * 
 * @param {Node} node A node with either children or nothing
 * 
 * @return {String} The binary in String format
 */
function serialize(node) {
   var s = ""

    if (node === undefined) {
        s += "# "
    }
    else {
        s += String(node.val) + " "
        s += serialize(node.left)
        s += serialize(node.right)
    }

    return s
}

/**
 * Deserialize a string into a binary tree. 
 * Nodes with empty children (Null) are marked with #
 * 
 * @param {List} stringTree A list of strings containing the tree
 * 
 * @return {Node} A binary tree 
 */
function deserialize(stringTree) {
    var val = stringTree.shift()
    if (val === '#') {
        return
    }
    else {
        node = new Node(val, deserialize(stringTree), deserialize(stringTree))
    }

    return node
 
}

var node = new Node('root', new Node('left', new Node('left.left')), new Node('right'))

var assert = require("assert")

assert(deserialize(serialize(node).split(" ")).left.left.val == 'left.left')
